### DISCUSSION TITLE
Support for bridging between javascript Set object and DynamoDB String/Number/Binary Set?

### DISCUSSION THREAD

**Discussion by tamakisquare** [sensedeep/dynamodb-onetable#67] [Q&A] [answered]
First of all, kudos to the team at SenseDeep for creating and sharing this wonderful library, which makes accessing and managing DynamoDB a much better experience.

After going through the official documentation and the code, it doesn't look like Set objects are being supported. Although I found a line of code documentation in _Model.js_ on Model `type` attribute mentioning Set type, the info is not only limited but also confusing (`Set` and `'Set'`?). 
<img width="844" alt="Screen Shot 2021-07-06 at 2 53 54 PM" src="https://user-images.githubusercontent.com/538353/124556119-7bcf1280-de6a-11eb-827d-f6bbec26dfb1.png">
Nevertheless, I still gave both `'Set'` and `Set` a try. 

When running with `type='Set'`, I got
> Error: Set value must be an array

Following the instruction in the error message, I further converted the Set object to its corresponding array object before writing to the db. This time around the value got written to the db successfully but it was stored as a List not a String Set like I wanted.

When running with `type=Set`, the Set object gets converted to an empty JS object, regardless the content of the Set, before writing to the db. 

I have also looked into the code to verify the observations mentioned above.

### Questions

1. OneTable currently does not support Set objects. Am I right?
2. If so, is there a plan or ongoing work to add such support to OneTable?

**Comment by mobsense** [comment]
Thank you for your kind words. Really appreciated.

Lists are supported, but sets are currently not supported. But we are working on that right now. Supporting nested schemas and "dot.qualified" updates just jumped ahead of them in the queue.

The issue they are not already implemented is that there isn't a transparent conversion from JavaScript objects to DDB sets and vice versa.  Lists and maps are easy as these are handled by the AWS SDK v2 DocumentClient and the v3 marshal APIs. But DocumentClient and V3 marshal don't do it automatically. We need to call CreateSet. 

We are experimenting and prototyping what would be the best way to specify sets and if we can just express them as lists of the appropriate type in Javascript and then have the schema specify the type as "set-string", "set-number" or "set-binary".

Do you have any suggestions?

**Reply by mobsense** [reply]
I expect set support to drop over the next week if it resolves smoothly.

**Reply by mobsense** [reply]
Progress update.

AWS SDK V3 is easy as the marshall / unmarshall handles sets.

The V2 SDK does not.  So we have to code that.

Onwards ....

**Comment by mobsense** [comment, marked as answer]
Okay, we've checked in Set support with test/data.ts which demonstrates basic use.

This is ready for test to see if it covers the bases. We'll continue to test, but would appreciate any parallel test you can offer.

This includes:

* Schema support for {type: Set}
* AWS SDK V2 and V3 support 
* Support for string, number and binary sets
* Update support via direct properties
* Update params.add, delete, set to modify Set elements

Basic operation is:

1. Define set types in the schema. The AWS SDKs interpolate the set type from the contents. Swimming against this seem difficult, so at this stage, we are defining the set type as "Set" regardless of whether it is a number, string or binary set.

```
stringSet:      { type: Set },
numberSet:   { type: Set },
binarySet:      { type: Set },
```

Then use the `Set` constructor to create values to provide to OneTable:

```
await Model.create({
    stringSet:  new Set(['one', 'two', 'three']),
    numberSet:  new Set([1, 2, 3]),
    binarySet:  new Set([Buffer.from('one'), Buffer.from('two'), Buffer.from('three')]),
})
```

Under the hood, SDK v2 does not handle de-wrapping sets back into JavaScript Set types whereas V3 does. So we patch the SDK v2 response to match V3.

**Reply by tamakisquare** [reply]
Thanks for expediting the support for Set types. 

For V3, I think it makes more sense to rely on the DynamoDBDocumentClient (past of the official SDK) as explained [here](https://github.com/sensedeep/dynamodb-onetable/discussions/68#discussioncomment-973759).

**Reply by mobsense** [reply]
But document client is just using the underlying marshall(), unmarshall() APIs which we do directly. i.e. one layer less code, but we use exactly the same marshalling code..

**Reply by tamakisquare** [reply]
@mobsense You are right. Functionally speaking, what DynamoDBDocumentClient does is the same as what OneTable has. They are essentially the same and thus redundant. From a maintenance perspective, it makes sense to consolidate the two to minimize wasted efforts. When the OneTable project started, DynamoDBDocumentClient was not available and thus the team had to include (un)marshalling code in the library. Now that DynamoDBDocumentClient is available and is production ready, I think it benefits the OneTable project by offloading this (un)marshalling part (for V3) to DynamoDBDocumentClient, so more maintenance efforts can be focused on the value-added parts of the OneTable project.
